src/main/scala/li/cil/oc/util/InventoryUtils.scala

```diff
@@ -61,40 +87,31 @@ object InventoryUtils {
    * The number of items inserted can be limited, to avoid unnecessary
    * changes to the inventory the stack may come from, for example.
    */
-  def insertIntoInventorySlot(stack: ItemStack, inventory: IInventory, side: Option[EnumFacing], slot: Int, limit: Int = 64, simulate: Boolean = false) =
+  def insertIntoInventorySlot(stack: ItemStack, inventory: IItemHandler, slot: Int, limit: Int = 64, simulate: Boolean = false): Boolean =
     (stack != null && limit > 0) && {
-      val isSideValidForSlot = (inventory, side) match {
-        case (inventory: ISidedInventory, Some(s)) => inventory.canInsertItem(slot, stack, s)
-        case _ => true
-      }
-      (stack.stackSize > 0 && inventory.isItemValidForSlot(slot, stack) && isSideValidForSlot) && {
-        val maxStackSize = math.min(inventory.getInventoryStackLimit, stack.getMaxStackSize)
-        val existing = inventory.getStackInSlot(slot)
-        val shouldMerge = existing != null && existing.stackSize < maxStackSize &&
-          existing.isItemEqual(stack) && ItemStack.areItemStackTagsEqual(existing, stack)
-        if (shouldMerge) {
-          val space = maxStackSize - existing.stackSize
-          val amount = math.min(space, math.min(stack.stackSize, limit))
-          stack.stackSize -= amount
-          if (simulate) amount > 0
-          else {
-            existing.stackSize += amount
-            inventory.markDirty()
-            true
-          }
+      val amount = math.min(stack.stackSize, limit)
+      if (simulate) {
+        val toInsert = stack.copy()
+        toInsert.stackSize = amount
+        inventory.insertItem(slot, toInsert, simulate) match {
+          case remaining: ItemStack => remaining.stackSize < stack.stackSize
```

Comment:
Shouldn't this be `remaining.stackSize < amount`? If `limit < stack.stackSize` nothing could have happened, and this would still be true, no?